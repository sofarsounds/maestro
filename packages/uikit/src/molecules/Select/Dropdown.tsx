import React, { useState, useRef } from 'react';

import Input from './Input';
import Menu from './Menu';
import { StickyContainer, PortalComponent } from '../../util/index';
import { useOutsideClick, useKeyDown } from '../../hooks';

interface SelectProps {
  children?: any;
  handleOptionClick?: (value: any) => void;
  readonly?: boolean;
  isOpen: boolean;
  component: React.ReactNode;
}

const Dropdown: React.SFC<SelectProps> = ({
  children,
  readonly,
  handleOptionClick,
  isOpen,
  component
}) => {
  const ref = useRef<any>();
  const [value, setValue] = useState<string>('');

  useOutsideClick(ref, () => {
    /* setIsOpen(false); */
  });

  useKeyDown('Escape', () => {
    /* setIsOpen(false); */
  });

  const optionClick = (value: any) => {
    setValue(value);
    /* setIsOpen(false); */
    handleOptionClick ? handleOptionClick(value) : null;
  };

  return (
    <>
      <PortalComponent dom={document.body}>
        <StickyContainer stickToEl={ref.current}>
          <Menu depth={3}>{children({ optionClick })}</Menu>
        </StickyContainer>
      </PortalComponent>
    </>
  );
};

export default Dropdown;
