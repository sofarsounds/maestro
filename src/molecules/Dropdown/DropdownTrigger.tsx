import styled from '../../lib/styledComponents';

export default styled.div`
  border: none;
  box-shadow: none;
  outline: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
  background-color: transparent;
`;
