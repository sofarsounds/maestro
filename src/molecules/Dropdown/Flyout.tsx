import styled, { css } from '../../lib/styledComponents';

import { withShadow } from '../../util/index';

export type FlyoutSizes = 'small' | 'large';

interface FlyoutProps {
  size?: FlyoutSizes;
  flyoutContainer?: boolean;
}

export default styled.div<FlyoutProps>`
  ${({ theme, size = 'regular', flyoutContainer = true }) => css`
    ${flyoutContainer &&
      css`
        background-color: ${theme.colors.whiteDenim};
        ${withShadow({ depth: 4 })};
        max-height: 300px;
        margin-top: ${theme.ruler[1]}px;
        border-radius: ${theme.borderRadius.large};
        padding: ${theme.ruler[8]}px ${theme.ruler[6]}px;

        overflow-y: auto;

        width: 200px;

        ${size === 'small' &&
          css`
            width: 150px;
          `}

        ${size === 'large' &&
          css`
            width: 250px;
          `}
      `}
  `}
`;
