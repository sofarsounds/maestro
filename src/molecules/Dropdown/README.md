# Dropdown

To implement a `Dropdown` into your project you'll need to add this import
```js
import { Dropdown } from '@sofarsounds/maestro'
```

After adding the import you can use the Dropdown simply like this
```html
<Dropdown label="Dropdown Label">
  I am the content!
</Dropdown>

<Dropdown renderLabel={() => <b>Custom <i>Label</i></b>}>
  I have a custom label
</Dropdown>
```

## Implements

- [useDisableScroll](../../hooks/useDisableScroll)

## Props

| Name                  | Type          | Default         | Description                      |
| :------------         | :-----        | :-------------- | :------------------------------- |
| **children**          | `React.Node`  |                 | The content displayed in the dropdown
| label                 | `string`      |                 | Label of the component
| renderLabel           | `Function`    |                 | Optional prop to render a React component as Label
| offsetTop             | `Number`      |                 | Moves the dropdown content down by given amount of pixels
| size                  | [Enum](#enum) |                 | The size of the flyout component
| flyoutContainer       | `Boolean`     |                 | Whether to display the default flyout container
| disableScrollWhenOpen | `Boolean`     |                 | Disables body scroll when dropdown is open
| keepInViewPort        | `Boolean`     |                 | Always keep the dropdown flyout within the viewport
| data-qaid             | `string`      |                 | Optional prop for testing purposes
| id                    | `string`      |                 | Default HTML id prop to identify the element

### Enum

| Size  |
| :---  |
| small |
| large |
