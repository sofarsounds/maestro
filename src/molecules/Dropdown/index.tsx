import React, { useState, useRef } from 'react';
import styled, { css } from '../../lib/styledComponents';

import Popper, { PopperProps } from '../../atoms/Popper';
import Portal from '../../atoms/Portal';
import Menu from '../../atoms/Menu';
import { useDisableScroll, useOutsideClick } from '../../hooks';

import Trigger from './DropdownTrigger';

interface DropdownProps extends PopperProps {
  label?: string;
  renderLabel?: (arg?: any) => any;
  children: any;
  size?: 'small' | 'large';
  flyoutContainer?: boolean; // TODO rename to `hasFlyoutContainer` ?
  disableScrollWhenOpen?: boolean;
  'data-qaid'?: string;
  id?: string;
}

export const StyledMenu = styled(Menu)`
  ${({ theme }) => css`
    padding: ${theme.ruler[8]}px ${theme.ruler[6]}px;
  `}
`;

const Dropdown: React.SFC<DropdownProps> = ({
  label,
  renderLabel,
  children,
  flyoutContainer = true,
  disableScrollWhenOpen = false,
  anchorOrigin = {
    vertical: 'bottom',
    horizontal: 'left'
  },
  transformOrigin = {
    vertical: 'top',
    horizontal: 'left'
  },
  offset,
  keepInViewPort,
  size,
  'data-qaid': qaId,
  id
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const ref = useRef<any>();
  const menuRef = useRef<any>();

  useDisableScroll(isOpen, disableScrollWhenOpen);

  useOutsideClick(menuRef, () => {
    setIsOpen(false);
  });

  let width = '200px';
  if (size === 'small') {
    width = '150px';
  } else if (size === 'large') {
    width = '250px';
  }

  return (
    <>
      <Trigger
        id={id}
        ref={ref}
        onClick={() => setIsOpen(!isOpen)}
        data-qaid={qaId}
      >
        {renderLabel ? renderLabel(isOpen) : label}
      </Trigger>

      {isOpen && (
        <Portal dom={document.body}>
          <Popper
            offset={offset}
            anchorEl={ref}
            anchorOrigin={anchorOrigin}
            transformOrigin={transformOrigin}
            keepInViewPort={keepInViewPort}
          >
            {flyoutContainer ? (
              <StyledMenu width={width} ref={menuRef} data-qaid={`${qaId}-flyout`}>
                {children}
              </StyledMenu>
            ) : (
              children
            )}
          </Popper>
        </Portal>
      )}
    </>
  );
};

export default Dropdown;
