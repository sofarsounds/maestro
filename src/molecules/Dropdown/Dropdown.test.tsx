import React from 'react';

import { mountWithTheme } from '../../test';
import Dropdown, { StyledMenu } from './index';
import Trigger from './DropdownTrigger';

const OutsideDiv = () => <button>bla</button>;

const setup = (flyoutContainer: boolean = true, offsetTop = 0) =>
  mountWithTheme(
    <>
      <OutsideDiv />
      <Dropdown
        flyoutContainer={flyoutContainer}
        label="I am a dropdown"
        offset={{ vertical: offsetTop }}
      >
        Dropdown Content
      </Dropdown>
    </>
  );

describe('<Dropdown />', () => {
  it('renders correctly', () => {
    expect(setup()).toMatchSnapshot();
  });

  it('renders the label correctly', () => {
    const wrapper = setup();
    expect(wrapper.find(Trigger).text()).toBe('I am a dropdown');
  });

  it('opens the dropdown flyout when the trigger is clicked', () => {
    const wrapper = setup();
    expect(wrapper.find(StyledMenu)).toHaveLength(0);
    wrapper.find(Trigger).simulate('click');
    expect(wrapper.find(StyledMenu)).toHaveLength(1);
  });

  it('renders the dropdown flyout content correctly', () => {
    const wrapper = setup();
    wrapper.find(Trigger).simulate('click');

    expect(wrapper.find(StyledMenu).text()).toBe('Dropdown Content');
  });

  it('closes the dropdown when clicking on the trigger again', () => {
    const wrapper = setup();
    wrapper.find(Trigger).simulate('click');
    wrapper.find(Trigger).simulate('click');
    expect(wrapper.find(StyledMenu)).toHaveLength(0);
  });

  it('render the flyout container with the correct styles', () => {
    const wrapper = setup(true);
    wrapper.find(Trigger).simulate('click');
    expect(wrapper.find(StyledMenu)).toHaveStyleRule('width', '200px');
  });

  it('does not render the styled flyout container when the prop is false', () => {
    const wrapper = setup(false);
    wrapper.find(Trigger).simulate('click');
    expect(wrapper.find(StyledMenu)).toHaveLength(1);
    expect(wrapper.find(StyledMenu)).not.toHaveStyleRule('width', '200px');
  });

  it.todo('closes the flyout when clicked outside');
  it.todo('closes the flyout when keydown escape');
});
