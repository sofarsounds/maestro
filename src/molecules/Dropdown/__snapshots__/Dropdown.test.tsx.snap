// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> renders correctly 1`] = `
.c0 {
  border: none;
  box-shadow: none;
  outline: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
  background-color: transparent;
  display: inline-block;
}

<Component
  theme={
    Object {
      "borderRadius": Object {
        "button": "4px",
        "default": "2px",
        "input": "2px",
        "large": "8px",
      },
      "colors": Object {
        "backToBlack": "#111111",
        "black": "#000000",
        "blackBetty": "#1E1E1E",
        "blueChristmas": "#3498DB",
        "blueMagic": "#34495E",
        "blueSmoke": "#949494",
        "brownSugar": "#9B525F",
        "checkBerry": "#98035A",
        "green100": "#CFEACC",
        "green200": "#9DDCAC",
        "green300": "#72D08A",
        "green400": "#4FC571",
        "green50": "#E6F7EA",
        "green500": "#23BA57",
        "green600": "#10AD52",
        "green700": "#039842",
        "green800": "#008736",
        "green900": "#006721",
        "greenGarden": "#19BC9B",
        "greenOnions": "#8BC34B",
        "macyGrey": "#DBDBDB",
        "mellowYellow": "#F1C410",
        "orangeCrush": "#F39C12",
        "paintItBlack": "#333333",
        "peaches": "#F99973",
        "pinkTriangle": "#E91E63",
        "primary": "#039842",
        "purpleHaze": "#3F51B5",
        "purplePills": "#9B59B6",
        "redRedWine": "#EF4423",
        "royalBlood": "#FF5000",
        "silverSprings": "#F5F5F5",
        "whiteDenim": "#FFFFFF",
        "yellowSubmarine": "#FFD426",
      },
      "dimensions": Object {
        "button": Object {
          "default": Object {
            "height": "44px",
            "padding": "0 25px",
          },
          "small": Object {
            "height": "40px",
            "padding": "0 15px",
          },
        },
        "buttonHeight": "44px",
        "containerWidth": Object {
          "default": "800px",
          "fullscreen": "100%",
          "large": "1000px",
          "small": "600px",
        },
        "inputHeight": "44px",
        "navbarHeight": Object {
          "lg": "84px",
          "md": "74px",
          "xs": "62px",
        },
        "radioBox": "20px",
      },
      "fontSizes": Object {
        "base": "16px",
        "body1": "16px",
        "body2": "14px",
        "button": "14px",
        "caption": "10px",
        "formGroupLabel": "14px",
        "h1": Object {
          "lg": "48px",
          "md": "34px",
          "xs": "24px",
        },
        "h2": Object {
          "lg": "34px",
          "md": "24px",
          "xs": "20px",
        },
        "h3": Object {
          "lg": "24px",
          "md": "20px",
          "xs": "16px",
        },
        "h4": Object {
          "lg": "20px",
          "md": "16px",
          "xs": "16px",
        },
        "h5": Object {
          "lg": "16px",
          "md": "14px",
          "xs": "14px",
        },
        "h6": Object {
          "lg": "14px",
          "md": "14px",
          "xs": "14px",
        },
        "overline": "10px",
        "title": Object {
          "lg": "54px",
          "md": "38px",
          "xs": "26px",
        },
      },
      "fonts": Object {
        "regular": "Open Sans",
      },
      "media": Object {
        "lg": [Function],
        "md": [Function],
        "sm": [Function],
        "xl": [Function],
        "xs": [Function],
      },
      "ruler": Array [
        0,
        4,
        8,
        12,
        16,
        20,
        24,
        28,
        32,
        36,
        40,
        44,
        48,
        52,
        56,
        60,
        64,
        68,
        72,
        76,
        80,
      ],
      "utils": Object {
        "transition": [Function],
      },
      "zIndex": Object {
        "navbar": 100,
      },
    }
  }
>
  <OutsideDiv>
    <button>
      bla
    </button>
  </OutsideDiv>
  <Dropdown
    flyoutContainer={true}
    label="I am a dropdown"
    offset={
      Object {
        "vertical": 0,
      }
    }
  >
    <styled.div
      onClick={[Function]}
    >
      <div
        className="c0"
        onClick={[Function]}
      >
        I am a dropdown
      </div>
    </styled.div>
  </Dropdown>
</Component>
`;
